File: source/downloads/code/LeetCode/Java/Number-of-Islands.java
public class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        if(m==0) return 0;
        int n = grid[0].length;
        int ret = 0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j] == '1'){
                    ret++;
                    dfs(grid,i,j);
                }
            }
        }
        return ret;
    }

    public void dfs(char[][] grid, int m, int n){
        if(m<0 || m>=grid.length || n<0 || n>=grid[0].length || grid[m][n]!='1')
            return ;
        grid[m][n] = '0';
        dfs(grid,m-1,n);
        dfs(grid,m+1,n);
        dfs(grid,m,n-1);
        dfs(grid,m,n+1);
    }
}

File: downloads/code/LeetCode/Python/Number-of-Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        row = len(grid)
        col = len(grid[0])
        ret = 0

        for i in range(row):
            for j in range(col):
                if grid[i][j] != '1':
                    continue
                ret += 1
                self.dfs(i,j,grid)
        return ret
    def dfs(self,i,j,grid):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] != '1':
            return
        grid[i][j] = '2'
        self.dfs(i-1,j,grid)
        self.dfs(i+1,j,grid)
        self.dfs(i,j-1,grid)
        self.dfs(i,j+1,grid)

